File: sampleproject/sampleproject/src/main/java/com/sampleproject/sampleproject/service/BookService.java
package com.sampleproject.sampleproject.service;

import com.sampleproject.sampleproject.dto.BookRequest;
import com.sampleproject.sampleproject.exception.ResourceNotFoundException;
import com.sampleproject.sampleproject.model.Book;
import com.sampleproject.sampleproject.repository.BookRepository;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class BookService {

    private final BookRepository bookRepository;

    public BookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    public Book createBook(BookRequest request) {
        Book book = new Book();
        book.setTitle(request.getTitle());
        book.setAuthor(request.getAuthor());
        book.setGenre(request.getGenre());
        book.setPages(request.getPages());
        book.setPublishedYear(request.getPublishedYear());
        return bookRepository.save(book);
    }

    public List<Book> getAllBooks(String author, String genre) {
        List<Book> books = bookRepository.findAll(Sort.by("id"));

        if (author != null) {
            books = books.stream()
                    .filter(b -> b.getAuthor().equalsIgnoreCase(author))
                    .collect(Collectors.toList());
        }
        if (genre != null) {
            books = books.stream()
                    .filter(b -> b.getGenre().equalsIgnoreCase(genre))
                    .collect(Collectors.toList());
        }
        return books;
    }

    public Book getBookById(Long id) {
        return bookRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Book not found with id: " + id));
    }
}

File: sampleproject/frontend/src/api/bookService.js
import axios from 'axios';

const API = axios.create({ baseURL: '/books' });

export async function getAllBooks(filters = {}) {
  const params = {};
  if (filters.author) params.author = filters.author;
  if (filters.genre) params.genre = filters.genre;
  const { data } = await API.get('', { params });
  return data;
}

export async function getBookById(id) {
  const { data } = await API.get(`/${id}`);
  return data;
}

export async function createBook(book) {
  const { data } = await API.post('', book);
  return data;
}
